### PR TITLE
[Loading pt.2] Style/Code update + better visual loading UX

### PR DESCRIPTION
**Pull request by W3stside**
Part of #735 

Note: some code/assets are not meant to be permanent choices, e.g the cow gif is funny and I like it but can be replaced. 

Improves loading UX/UI
Screenshot 2021-06-03 at 11 31 38

Loader now in price bar, as compared to previous swap button placement

### CODE REVIEW

File: src/custom/components/swap/TradePrice/index.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    width: 30px;
+    height: 30px;
+    position: relative;
+    overflow: hidden;
+  }
+`
+
+const LoadingCowComponent = ({ showCow }: { showCow: boolean }) => (
+  <>
+    <AnimatedImg src={loadingCowGif} showLoader={showCow} title="Prices updating..." />
```
Comment (alfetopito) [inline src/custom/components/swap/TradePrice/index.tsx:22]: I think the other way around feels better:

```suggestion
    <AnimatedImg src={loadingCowGif} showLoader={showCow} title="Updating prices..." />
```

File: src/custom/components/swap/TradePrice/index.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    <AnimatedImg src={repeat} showLoader={!showCow} width="30px" padding="6px" />
+  </>
+)
+
+export default function TradePrice(props: TradePriceProps) {
+  const isQuoteSoftLoading = useIsQuoteSoftLoading()
+  const showCow = useLoadingWithTimeout(isQuoteSoftLoading, 4000)
```
Comment (alfetopito) [inline src/custom/components/swap/TradePrice/index.tsx:29]: I think we can have a smaller timeout.
2s might be enough.

Also, magic number :sparkles:

File: src/custom/hooks/useLoadingWithTimeout.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  const [delayedLoad, setDelayedLoad] = useState(isLoading)
+
+  useEffect(() => {
+    let timeout: NodeJS.Timeout | undefined
+    // price is being queried
+    if (isLoading) {
+      // isn't currently showing cow or hasnt yet been updated here
+      // so we clear any running timeouts ready to clear local loading state
+      // and essentially reset them
+      clearTimeout(timeout as NodeJS.Timeout)
```
Comment (alfetopito) [inline src/custom/hooks/useLoadingWithTimeout.ts:13]: Shouldn't you clear only if it's set?
What happens with this cast when it's `undefined`?

```suggestion
      if (timeout) {
        clearTimeout(timeout)
      }
```

Comment (alfetopito) [reply src/custom/hooks/useLoadingWithTimeout.ts:13]: Same for all timeouts here

**Review by alfetopito — COMMENTED** [review]
I absolutely love seeing the nose picking cow. Makes me happy every time :joy: 

One thing that still feels weird though, is that when I change parameters, it briefly shows "insufficient liquidity" before showing the updated price.

I would expect the ellipsis gif to be displayed in the button instead of the default liquidity message.

Otherwise, some minor comments

**Comment by W3stside** [comment]
Updated. logic should be added to base PR but styles will change

File: src/custom/hooks/useLoadingWithTimeout.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    // price is being queried
+    if (isLoading) {
+      // isn't currently showing cow or hasnt yet been updated here
+      // so we clear any running timeouts ready to clear local loading state
+      // and essentially reset them
+      timeout && clearTimeout(timeout)
+      setDelayedLoad(true)
+    } else {
+      // no longer loading
+      // reset timeout to clear local loading state after LOADING_COW_TIMER ms
```
Comment (alfetopito) [inline src/custom/hooks/useLoadingWithTimeout.ts:17]:
```suggestion
      // reset timeout to clear local loading state after `time` ms
```

File: src/custom/hooks/useLoadingWithTimeout.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      if (delayedLoad) {
+        timeout = setTimeout(() => {
+          clearTimeout(timeout as NodeJS.Timeout)
```
Comment (alfetopito) [inline src/custom/hooks/useLoadingWithTimeout.ts:20]: You forgot one :)

```suggestion
          timeout && clearTimeout(timeout)
```

File: src/custom/components/swap/TradePrice/index.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+const TradePriceMod = styled(TradePriceModComponent)`
+  ${StyledBalanceMaxMini} {
+    width: 30px;
+    height: 30px;
+    position: relative;
+    overflow: hidden;
+  }
+`
+const COW_LOADING_TIME = 4000
```
Comment (alfetopito) [inline src/custom/components/swap/TradePrice/index.tsx:19]: I still think 4s is way too long.
We don't have to wait until it's done, but to me the UI indicates it's not ready until the nose-picking-cow is gone.

**Review by alfetopito — APPROVED** [review]
Mostly what I said before.

It looks great, I love the cow, and I still see insufficient liquidity when updating the amounts.
In case that's not supposed to be addressed here, it's good to go.

screenshot_2021-06-08_07-40-24

Approved with minor comments/suggestions

File: src/custom/components/swap/TradePrice/index.tsx (hunk 6 above)
Comment (W3stside) [reply src/custom/components/swap/TradePrice/index.tsx:19]: maybe I'll go with @anxolin 's siuggestion from another pr then

**Review by anxolin — APPROVED** [review]

**Pull request merged**